ls1intum/Artemis
# Enhancement/better result feedback

### Checklist

### Motivation and Context


There are some minor UI improvements we wanted to make. Basically as of now, we almost always have `java.lang.AssertionError: ` as a prefix for every failed test result feedback for Java programming exercises. This doesn't add any informational value to the feedback.
We also shouldn't split the lines of the test output for C/Python tests, since it makes sense in these cases to show the whole output. The reason being that the whole feedback contains useful information in these cases.

There also was a bug: When fetching all participations for a student ID (for all active exercises), the in the background built SQL is syntactically incorrect if the set of active exercises is empty --> `...where foo is in ())` (Note the empty brackets at the end, normally there is a set of exerciseIds). 

### Description


1. Show test results for programming exercises in the detail view in the top right corner in addition to the bottom of the page.
2. Don't show `java.lang.AssertionError` in the Java feedback
3. Don't split up the feedback for C/Python exercises.
4. Fix the above described bug by not executing the request at all for empty exercise sets.

For already existing exercise feedback, I could retroactively filter out the unwanted AssertionError strings, but we would have to somehow hardcode the ID of "older" exercises in order to decide when to apply the filter. I think the best approach here would be to just run an SQL script, which modifies all feedback strings and remove unwanted parts:
```
update feedback
set detail_text = replace(detail_text, 'java.lang.AssertionError: ', '')
where id in (
    select * from (select f.id
    from feedback f
             join result r on f.result_id = r.id
             join participation p on r.participation_id = p.id
             join exercise e on p.exercise_id = e.id
    where e.programming_language = 'JAVA'
)frpei);
```

### Steps for Testing


1. Log in to Artemis
2. Navigate to any course
3. Open a programming exercise in the code editor
4. Submit code with errors
5. The feedback shouldn't contain any AssertionError strings
6. Open the detail view for the same exercise
7. Your results should also be displayed in the top right corner.

### Screenshots


![result_detail_top](https://user-images.githubusercontent.com/27979674/64197417-66f85100-ce86-11e9-9f75-7e163a2e678c.png)
![test_names](https://user-images.githubusercontent.com/27979674/64493278-7d375000-d27e-11e9-9dfb-8e6cd6b0da8e.png)
![no_test_names](https://user-images.githubusercontent.com/27979674/64493275-714b8e00-d27e-11e9-855a-2d13d2dbb706.png)


![image](https://user-images.githubusercontent.com/28230611/64206615-1e975e00-ce9b-11e9-8837-5c10c6e2a304.png)
